src/gui/macOS/fileproviderdomainmanager.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2022 by Claudio Cambra <claudio.cambra@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#pragma once
+
+#include <QObject>
+
+#include "accountstate.h"
+
+namespace OCC {
+
+class Account;
+
+namespace Mac {
+
+class FileProviderDomainManager : public QObject
+{
+    Q_OBJECT
+
+public:
+    explicit FileProviderDomainManager(QObject * const parent = nullptr);
+    ~FileProviderDomainManager() override;
+
+private slots:
+    void setupFileProviderDomains();
+
+    void addFileProviderDomainForAccount(const OCC::AccountState * const accountState);
+    void removeFileProviderDomainForAccount(const OCC::AccountState * const accountState);
+    void disconnectFileProviderDomainForAccount(const OCC::AccountState * const accountState, const QString &reason);
+    void reconnectFileProviderDomainForAccount(const OCC::AccountState * const accountState);
+
+    void trySetupPushNotificationsForAccount(const OCC::Account * const account);
+    void setupPushNotificationsForAccount(const OCC::Account * const account);
+    void signalEnumeratorChanged(const OCC::Account * const account);
+
+    void slotAccountStateChanged(const OCC::AccountState * const accountState);
+    void slotEnumeratorSignallingTimerTimeout();
+
+private:
+    // Starts regular enumerator signalling if no push notifications available
+    QTimer _enumeratorSignallingTimer;
+
+    class Private;
+    std::unique_ptr<Private> d;
```

Comment:
please do not do private implementation via `d` pointer
we do not care about binary compatibility or similar considerations

Reply:
If we don't do this then our code becomes a real mess in the Objective-C++ side;

as a general practice I have tried to separate the C++ from the Objective-C code as much as possible in our Objective-C++ code, a private implementation is the cleanest way to do this

Reply:
> the mix of languages is already the case

No. Including Objective-C code in the header would break compilation on the C++ side, so there is a reason we have the implementation fully hidden away inside a "private" class and not exposed at all in the header. 

Having a private implementation is a way to continue this separation into the actual implementation file -- the more we partition C++ code from Objective-C, the less we will run into the sharp edges of c++/objective-c comingling and the more maintainable the code will be

Reply:
> that said, maybe `Private` is not the best name considering the reason for the existence of this class
> `MacFileProviderDomainManager` may be a better solution (or `FileProviderDomainManagerPrivate` but that one is very long)

Okay, good idea -- renamed to FileProviderDomainManager::MacImplementation